rust/datafusion/src/physical_plan/union.rs

```diff
@@ -60,15 +60,31 @@ impl ExecutionPlan for UnionExec {
 
     /// Output of the union is the combination of all output partitions of the inputs
     fn output_partitioning(&self) -> Partitioning {
-        // Sums all the output partitions
-        let num_partitions = self
-            .inputs
+        let intial: Option<Partitioning> = None;
+        self.inputs
             .iter()
-            .map(|plan| plan.output_partitioning().partition_count())
-            .sum();
-        // TODO: this loses partitioning info in case of same partitioning scheme (for example `Partitioning::Hash`)
-        // https://issues.apache.org/jira/browse/ARROW-11991
-        Partitioning::UnknownPartitioning(num_partitions)
+            .fold(intial, |acc, input| {
+                match (acc, input.output_partitioning()) {
+                    (None, partition) => Some(partition),
+                    (
+                        Some(Partitioning::Hash(mut vector_acc, size_acc)),
+                        Partitioning::Hash(vector, size),
+                    ) => {
+                        vector_acc.append(&mut vector.clone());
```

Comment:
I am not sure this is correct -- of the input is hash partitioned by some expression across `size_acc` partitions, the output will only be hash partitioned if the hash expressions are the same and the number of output partitions is the same, right?

Reply:
I think so too. Maybe just look for the first input (there is always at least 2 in union) and compare with the others to have same expression and same nr of partitions.

Reply:
Just to make sure I am on the same page: 

Two `Partitioning::Hash` may only be composed if they have the same expression and number of partitions.  If the two `Partitioning::Hash` differ in expression and/or number of partitions then it becomes a `UnknownPartitioning`. Correct?

Reply:
That is my understanding

Reply:
@alamb I am having trouble figuring out how to check for equality between `PhysicalExpr`.  Do I need to implement a `PartialEq` or is there a more clever way?

Reply:
🤔  I suspect we will need to implement `PartialEq` (which you may be able to do via `#[derive(PartialEq)]` annotation on `PhysicalExpr`

Reply:
I am still new to rust, so sorry for the beginner questions.  Thank you for your patience.

Since `PhysicalExpr` is a trait and not a struct or enum I am unable to use `#[derive(PartialEq)]`.  Additionally I am not able to do the following: `pub trait PhysicalExpr: Send + Sync + Display + Debug  + PartialEq`.  My guess is for why this doesn't work is the same reason the `derive` isn't working, which is that `PhysicalExpr` isn't a struct or enum.  Do you know how to require all objects that implement a trait also implement `PatialEq` or is there a different way?

Reply:
I would have expected

```
pub trait PhysicalExpr: Send + Sync + Display + Debug + PartialEq
```

To be the correct way as it says that everything that implements `PhysicalExpr` also needs to implement `PartialEq`. You would then  have to `#[derive(PartialEq)]` or `impl PartialEq` for all structs that claim to `impl PhysicalExpr`

What error did you get when you tried that?

Reply:
So I added `PartialEq` and ensured all structs that impl `PhysicalExpr` also `impl PartialEq`.  For every occurrence of `Arc<dyn PhysicalExpr>` I get the following error, 132 times:

```bash
error[E0038]: the trait `PhysicalExpr` cannot be made into an object
   --> datafusion/src/physical_plan/udf.rs:95:17
    |
95  |     args: &[Arc<dyn PhysicalExpr>],
    |                 ^^^^^^^^^^^^^^^^ `PhysicalExpr` cannot be made into an object
    |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> datafusion/src/physical_plan/mod.rs:188:57
    |
188 | pub trait PhysicalExpr: Send + Sync + Display + Debug + PartialEq {
    |           ------------                                  ^^^^^^^^^ ...because it uses `Self` as a type parameter
    |           |
    |           this trait cannot be made into an object...

```

It appears that the inclusion of PartialEq makes it so `PhysicalExpr` doesn't play nicely with dynamic dispatch.  Due to PartialEq taking a type arg of `Self`, which can't be determined under dynamic dispatch.  I am not sure how to work around this.

Reply:
🤔 it seems [stackoverflow](https://stackoverflow.com/questions/25339603/how-to-test-for-equality-between-trait-objects) has some context

Perhap we need to implement our own version of equality for `PhysicalExpr` or something.

I am not sure how to proceed, to be honest. @Dandandan  any thoughts?